janus_core/calculations/phonons.py

```diff
@@ -8,6 +8,7 @@
 import phonopy
 from phonopy.file_IO import write_force_constants_to_hdf5
 from phonopy.structure.atoms import PhonopyAtoms
+from tqdm import tqdm
```

Comment:
`tqdm` may be a dependency of a dependency and that's not a good thing to rely on. You need to add this explicitly to dependencies if you want to use it.

Reply:
Also, `rich` is also a dependency of a dependency, that *may* be a better choice here since Jupyter notebooks and IPython are encouraged ways of using Janus

Reply:
Do you know how rich compares in terms of performance/overheads?

Reply:
I have no idea, but I should think minimal in relation to a phonon calculation.

Reply:
[Supposedly](https://github.com/Textualize/rich/issues/176#issuecomment-663995614), they should be pretty comparable nowadays. Regardless, as mentioned, the milliseconds are going to be negligible.

Reply:
I think in the context of issues linked in #272, some of the slowness of typer seemed to be due to rich imports on timescales noticeable by users, but I'm happy with which ever seems nicer, as long as performance seems ok

Reply:
Oof, I don't know about imports, to be honest. `tqdm` is supposed to be leightweight and has no dependencies, so it might be better in that regard, but then again this implementation is in `calculations`, so as long as the import of the `calculations.phonons` happens after the CLI is loaded (which is what that PR proposes, if I'm not wrong), then it shouldn't make a difference, right?

Reply:
If `rich` is already imported due to other imports, it will already be in the loaded modules and thus free to import.

Reply:
> Oof, I don't know about imports, to be honest. `tqdm` is supposed to be leightweight and has no dependencies, so it might be better in that regard, but then again this implementation is in `calculations`, so as long as the import of the `calculations.phonons` happens after the CLI is loaded (which is what that PR proposes, if I'm not wrong), then it shouldn't make a difference, right?

Sorry, I meant just in terms of it potentially having a bigger impact than you might expect when running calculations.

> If `rich` is already imported due to other imports, it will already be in the loaded modules and thus free to import.

True, I think it will imported if you're using the CLI, but not necessarily if you're using the Python interface.

I think it should be fine, if it is nicer.

Reply:
Personally, I prefer the `tqdm` default progress bar, though `rich` appears to be more customisable

Reply:
That's all fine if everyone's happy. The main thing was that it was an option which is why (although not necessarily obvious in the GitHub font) "may" was italicised.

Reply:
> On the other hand, it does appear to me that `tqdm` integration with jupyter notebook could be non-trivial

Don't worry the integration is that it prints to the output perfectly happily.

Reply:
So, to confirm, we are sticking with `tqdm`, is that right?

Reply:
I obviously haven't seen how this specific one would look in comparison, but in general `rich` maybe looks slightly nicer to me, and maybe fits stylistically with the typer interface better.

So, all things being equal (e.g. no immediately noticeable performance issues, which it sounds like there shouldn't be), I'd maybe tend towards `rich`.

I am completely fine with `tqdm` if it's more straightforward to finish this PR up though, as long as it addresses the issue.

Reply:
This is how the different implementations look on my linux VM:

`tqdm`:
in progress:
![image](https://github.com/user-attachments/assets/ca209005-7690-41ca-8300-0550e54e51d9)

complete:
![image](https://github.com/user-attachments/assets/745045e7-6dc7-49cb-bb2c-bd492df00a1c)

code:
```
        if self.enable_progress_bar:
            disp_supercells = tqdm(disp_supercells, desc="Computing displacements", unit="displacement")

        phonon.forces = [
            self._calc_forces(supercell)
            for supercell in disp_supercells
            if supercell is not None
        ]

        if self.enable_progress_bar:
            disp_supercells.write("Displacements complete; computing requested properties...")
```

`rich` basic useage:
in progress:
![image](https://github.com/user-attachments/assets/e6ceb3a1-513c-4b65-a9f4-5d785d393958)
complete:
![image](https://github.com/user-attachments/assets/e1d34915-0a41-4280-b2eb-d65cabe73197)
code:
```
        if self.enable_progress_bar:
            disp_supercells = track(disp_supercells, description="Computing displacements...")

        phonon.forces = [
            self._calc_forces(supercell)
            for supercell in disp_supercells
            if supercell is not None
        ]
```

`rich` with customisations:
in progress:
![image](https://github.com/user-attachments/assets/b8f006fe-7faa-4517-89eb-289ab41215c1)

complete:
![image](https://github.com/user-attachments/assets/462bd7a2-d2b3-4c5d-824f-287212932ab8)

code:
```
        if self.enable_progress_bar:
            text_column = TextColumn("{task.description}")
            bar_column = BarColumn(bar_width=None)
            completion_column = MofNCompleteColumn()
            time_column = TimeRemainingColumn()
            progress = Progress(text_column, bar_column, completion_column, time_column, expand=True)
            with progress:
                disp_supercells = progress.track(disp_supercells, description="Running displacements...")
                phonon.forces = [
                        self._calc_forces(supercell)
                        for supercell in disp_supercells
                        if supercell is not None
                ]
        else:
            phonon.forces = [
                self._calc_forces(supercell)
                for supercell in disp_supercells
                if supercell is not None
            ]
```

Reply:
They both look nice to me! Comes down to what Janus is likely to use more widely.

(One other program that could use a progress bar is `singlepoint`, when applied to a file containing many structures.)

Reply:
Thanks, @RastislavTuranyi!

On balance, I think I'd prefer rich. I don't mind too much about customisations, as long as the options are sensible.

> (One other program that could use a progress bar is `singlepoint`, when applied to a file containing many structures.)

I suppose it depends how often people are using them for large numbers of structures, since each individual calculation is generally pretty quick. (Descriptors can also be calculated for a trajectory, so could also benefit)

Reply:
> > (One other program that could use a progress bar is `singlepoint`, when applied to a file containing many structures.)
> 
> I suppose it depends how often people are using them for large numbers of structures, since each individual calculation is generally pretty quick. (Descriptors can also be calculated for a trajectory, so could also benefit)

It's quite handy as part of a workflow where configurations are sampled. Getting the results in a single file is also useful in such cases!

Reply:
As long as your input is a single file with many configurations, the output of `singlepoint` should also be a single file with the results as part of the info for each configuration, or do you mean in other cases? 

(I'm going to open an issue for potentially using the progress bar elsewhere, and resolve this since the original comment has been addressed)

Reply:
Yes, I think this is the existing (and desired) behaviour